tests/mesh/mesh_function_dfem.C

```diff
@@ -0,0 +1,218 @@
+public:
+  CPPUNIT_TEST_SUITE( MeshfunctionDFEM );
+
+  CPPUNIT_TEST( test_point_locator_dfem );
+
+  CPPUNIT_TEST( test_mesh_function_dfem );
+
+  CPPUNIT_TEST_SUITE_END();
+
+protected:
+  SerialMesh * _mesh;
+  UniquePtr<PointLocatorBase> _point_locator;
+
+  void build_mesh()
+  {
+    _mesh = new SerialMesh(*TestCommWorld);
+
+    /*
+      (0,1)           (1,1)
+        x---------------x
+        |               |
+        |               |
+        |               |
+        |               |
+        |               |
+        x---------------x
+       (0,0)           (1,0)
+        |               |
+        |               |
+        |               |
+        |               |
+        x---------------x
+       (0,-1)          (1,-1)
+     */
+
+    _mesh->set_mesh_dimension(2);
+
+    _mesh->add_point( Point(0.0,-1.0), 4 );
+    _mesh->add_point( Point(1.0,-1.0), 5 );
+    _mesh->add_point( Point(1.0, 0.0), 1 );
+    _mesh->add_point( Point(1.0, 1.0), 2 );
+    _mesh->add_point( Point(0.0, 1.0), 3 );
+    _mesh->add_point( Point(0.0, 0.0), 0 );
+
+    {
+      Elem* elem_top = _mesh->add_elem( new Quad4 );
+      elem_top->set_node(0) = _mesh->node_ptr(0);
+      elem_top->set_node(1) = _mesh->node_ptr(1);
+      elem_top->set_node(2) = _mesh->node_ptr(2);
+      elem_top->set_node(3) = _mesh->node_ptr(3);
+
+      Elem* elem_bottom = _mesh->add_elem( new Quad4 );
+      elem_bottom->set_node(0) = _mesh->node_ptr(4);
+      elem_bottom->set_node(1) = _mesh->node_ptr(5);
+      elem_bottom->set_node(2) = _mesh->node_ptr(1);
+      elem_bottom->set_node(3) = _mesh->node_ptr(0);
+    }
+
+    // libMesh will renumber, but we numbered according to its scheme
+    // anyway. We do this because when we call uniformly_refine subsequenly,
+    // it's going use skip_renumber=false.
+    _mesh->prepare_for_use(false /*skip_renumber*/);
+
+    // get a point locator
+    _point_locator = _mesh->sub_point_locator();
+  }
+
+public:
+  void setUp()
+  {
+    this->build_mesh();
+  }
+
+  void tearDown()
+  {
+    delete _mesh;
+  }
+
+  // test that point locator works correctly
+  void test_point_locator_dfem()
+  {
+    // this point is in bottom element only
+    Point interior(0.5, -0.5, 0.0);
+
+    // this point is on the face between bottom and top
+    Point face(0.5, 0.0, 0.0);
+
+    // test interior point
+    std::set<const Elem *> int_cand;
+    (*_point_locator)(interior, int_cand, libmesh_nullptr);
+    CPPUNIT_ASSERT (int_cand.size() == 1);
+    for (std::set<const Elem *>::iterator it = int_cand.begin(); it != int_cand.end(); ++it)
+      CPPUNIT_ASSERT ((*it)->id() == 1);
+
+    // test interior point
+    std::set<const Elem *> face_cand;
+    (*_point_locator)(face, face_cand, libmesh_nullptr);
+    CPPUNIT_ASSERT (face_cand.size() == 2);
+    unsigned int count = 0;
+    for (std::set<const Elem *>::iterator it = face_cand.begin(); it != face_cand.end(); ++it)
+    {
+      // FIXME: this tests occasionally fails. Why?
```

Comment:
Since the std::set is sorted on pointers, you have no guarantee that Element 0's address is going to come before Element 1's address in memory.  So I'd change the check here to something else... maybe just
```
int array[2] = {0,0};
for (each face_cand)
  array[(*it)->id()] = 1;
// Check that the array is all 1's
```